File: python/thermo/bridge.py
"""
Data fusion bridge for thermo CLI.

Handles spawning subprocesses (like cmg-cli) and injecting
thermal data into their output streams.
"""

import json
import subprocess
import sys
from typing import Any

from .hardware import ThermoBoard


class FuseBridge:
    """
    Bridge for fusing thermal data into 'cmg-cli get' command output.

    Runs 'cmg-cli get' command, captures its output, and injects
    thermal readings from configured MCC 134 sources.
    """

    def __init__(self, sources: list[dict], get_args: list[str]):
        """
        Initialize the fuse bridge.

        Args:
            sources: List of source configs with 'address', 'channel', and optional 'key'
            get_args: Arguments to pass to 'cmg-cli get' command
        """
        self.sources = sources
        self.get_args = get_args
        self._boards: dict[int, ThermoBoard] = {}

    def _get_board(self, address: int) -> ThermoBoard:
        """Get or create a ThermoBoard instance for the given address."""
        if address not in self._boards:
            self._boards[address] = ThermoBoard(address)
        return self._boards[address]

    def _get_thermal_data(self) -> dict[str, float]:
        """
        Read all configured thermal sources.

        Returns:
            Dictionary mapping keys to temperature values
        """
        data = {}
        for src in self.sources:
            addr = src['address']
            ch = src['channel']
            tc_type = src['tc_type']
            key = src.get('key', f'TEMP_{addr}_{ch}')

            try:
                board = self._get_board(addr)
                board.set_tc_type(ch, tc_type)
                val = board.get_reading(ch, 'temp')
                data[key] = val
            except Exception as e:
                # On read error, inject NaN or skip
                data[key] = float('nan')

        return data
    
    def _inject_json(self, json_obj: dict, thermal_data: dict) -> dict:
        """Inject thermal data into a JSON object."""
        json_obj.update(thermal_data)
        return json_obj

    def _inject_plain_text(self, lines: list[str], thermal_data: dict) -> list[str]:
        """Inject thermal data into a plain text object."""
        # Calculate formatting of received lines:
        # KEY:     XX.XXXXXX UNIT
        max_key_length = 0
        max_int_length = 0
        max_frac_length = 0

        data: list[tuple[str, float, str]] = []
        for line in lines:
            semicolon_index = line.find(':')
            point_index = line.find('.')

            unit_space_index = line.find(' ', point_index)
            if unit_space_index == -1:
                unit_space_index = len(line)
                unit = ''
            else:
                unit = line[unit_space_index + 1:].strip()

            key = line[:semicolon_index]
            value_str = line[semicolon_index + 1:unit_space_index].strip()
            int_length = value_str.find('.')
            frac_length = len(value_str) - int_length - 1

            max_key_length = max(max_key_length, len(key))
            max_int_length = max(max_int_length, int_length)
            max_frac_length = max(max_frac_length, frac_length)

            value = float(value_str)
            data.append((key, value, unit))
        
        for key, value in thermal_data.items():
            value_str = str(value)
            max_key_length = max(max_key_length, len(key))
            max_int_length = max(max_int_length, len(value_str.split('.')[0]))
            data.append((key, value, 'degC'))
        
        format_string = f"{{:<{max_key_length + 1}}}  {{:{max_int_length + max_frac_length + 1}.{max_frac_length}f}} {{}}"
        injected_lines = []
        for key, value, unit in data:
            new_line = format_string.format(key + ':', value, unit)
            injected_lines.append(new_line)
        return injected_lines

    def run(self) -> int:
        """
        Run the subprocess and fuse thermal data into its output.

        Returns:
            Exit code from the subprocess
        """
        process = subprocess.Popen(
            ["stdbuf", "-oL", "-eL", "cmg-cli", "get"] + self.get_args,
            stdout=subprocess.PIPE,
            stderr=sys.stderr,  # Pass stderr through directly
            text=True,
            bufsize=1,  # Line buffered
        )

        is_json = True
        received_data = []
        received_keys = set[str]()

        try:
            for line in process.stdout:
                line = line.strip()
                if not line:
                    continue

                if ":" not in line:
                    print(line, flush=True)
                    continue

                # Fetch latest thermal data
                thermal_data = self._get_thermal_data()

                # Attempt to parse line as JSON (cmg-cli --json mode)
                if is_json:
                    try:
                        json_obj = json.loads(line)
                        # Inject thermal data at root level
                        self._inject_json(json_obj, thermal_data)
                        print(json.dumps(json_obj), flush=True)
                    except json.JSONDecodeError:
                        is_json = False

                if not is_json:
                    # Plain text mode - print original line then thermal data
                    key = line.split(':')[0]
                    if key in received_keys:
                        injected_lines = self._inject_plain_text(received_data, thermal_data)
                        for line in injected_lines:
                            print(line, flush=True)
                        received_data = []
                        received_keys = set[str]()

                    received_data.append(line)
                    received_keys.add(key)
                
            if received_data:
                injected_lines = self._inject_plain_text(received_data, thermal_data)
                for line in injected_lines:
                    print(line, flush=True)

        except KeyboardInterrupt:
            process.terminate()
            process.wait()
            return 130  # Standard exit code for SIGINT

        process.wait()
        return process.returncode

File: src/bridge.c
/*
 * Data fusion bridge for fusing thermal data into cmg-cli output.
 * Spawns cmg-cli as subprocess and injects thermal readings.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/wait.h>
#include <sys/time.h>
#include <time.h>
#include <getopt.h>
#include <daqhats/daqhats.h>

#include "bridge.h"
#include "hardware.h"
#include "common.h"
#include "signals.h"
#include "board_manager.h"

#include "cJSON.h"

struct FuseBridge {
    ThermalSource *sources;
    int source_count;
    char **args;
    int arg_count;
    BoardManager board_mgr;
    int boards_initialized;
    char time_format[64];
};

/* Create a new bridge instance */
FuseBridge* bridge_create(ThermalSource *sources, int source_count, char **args, int arg_count, const char *time_format) {
    FuseBridge *bridge = (FuseBridge*)malloc(sizeof(FuseBridge));
    
    bridge->sources = (ThermalSource*)malloc(source_count * sizeof(ThermalSource));
    memcpy(bridge->sources, sources, source_count * sizeof(ThermalSource));
    bridge->source_count = source_count;
    
    bridge->args = (char**)malloc(arg_count * sizeof(char*));
    for (int i = 0; i < arg_count; i++) {
        bridge->args[i] = strdup(args[i]);
    }
    bridge->arg_count = arg_count;
    bridge->boards_initialized = 0;
    strncpy(bridge->time_format, time_format, sizeof(bridge->time_format) - 1);
    bridge->time_format[sizeof(bridge->time_format) - 1] = '\0';
    
    return bridge;
}

/* Free bridge resources */
void bridge_free(FuseBridge *bridge) {
    if (!bridge) return;
    
    /* Close any open boards via BoardManager */
    if (bridge->boards_initialized) {
        board_manager_close(&bridge->board_mgr);
    }
    
    if (bridge->sources) free(bridge->sources);
    
    for (int i = 0; i < bridge->arg_count; i++) {
        free(bridge->args[i]);
    }
    free(bridge->args);
    free(bridge);
}

/* Initialize boards for continuous reading using BoardManager */
static int bridge_init_boards(FuseBridge *bridge) {
    /* Initialize BoardManager and open all boards */
    if (board_manager_init(&bridge->board_mgr, bridge->sources, bridge->source_count) != THERMO_SUCCESS) {
        return -1;
    }
    
    /* Configure calibration and TC types */
    board_manager_configure(&bridge->board_mgr);
    
    bridge->boards_initialized = 1;
    
    thermo_wait_for_readings();
    
    return 0;
}

/* Get thermal data from all configured sources (boards must be initialized) */
static cJSON* get_thermal_data(FuseBridge *bridge) {
    cJSON *data = cJSON_CreateObject();
    
    for (int i = 0; i < bridge->source_count; i++) {
        ThermalSource *src = &bridge->sources[i];
        double temp, adc, cjc;
        
        /* Create a sub-object for each source */
        cJSON *source_data = cJSON_CreateObject();
        
        /* Read temperature (board is already open with TC type set) */
        int temp_result = mcc134_t_in_read(src->address, src->channel, &temp);
        if (temp_result == RESULT_SUCCESS) {
            cJSON_AddNumberToObject(source_data, "TEMP", temp);
        } else {
            cJSON_AddNumberToObject(source_data, "TEMP", 0.0/0.0);
        }
        
        /* Read ADC voltage */
        int adc_result = mcc134_a_in_read(src->address, src->channel, OPTS_DEFAULT, &adc);
        if (adc_result == RESULT_SUCCESS) {
            cJSON_AddNumberToObject(source_data, "ADC", adc);
        } else {
            cJSON_AddNumberToObject(source_data, "ADC", 0.0/0.0);
        }
        
        /* Read CJC temperature */
        int cjc_result = mcc134_cjc_read(src->address, src->channel, &cjc);
        if (cjc_result == RESULT_SUCCESS) {
            cJSON_AddNumberToObject(source_data, "CJC", cjc);
        } else {
            cJSON_AddNumberToObject(source_data, "CJC", 0.0/0.0);
        }
        
        /* Add the source data object to the main data object */
        cJSON_AddItemToObject(data, src->key, source_data);
    }
    
    return data;
}

/*
 * Format timestamp with microsecond support.
 * Use %f in format string for 6-digit microseconds.
 */
static void format_timestamp(char *buf, size_t buf_size, const struct timeval *tv, const char *format) {
    struct tm *tm_info = localtime(&tv->tv_sec);
    
    /* First pass: replace %f with microseconds placeholder */
    char temp_format[128];
    char *dst = temp_format;
    const char *src = format;
    char usec_str[8];
    snprintf(usec_str, sizeof(usec_str), "%06ld", (long)tv->tv_usec);
    
    while (*src && (dst - temp_format) < (int)sizeof(temp_format) - 7) {
        if (*src == '%' && *(src + 1) == 'f') {
            /* Replace %f with microseconds */
            strcpy(dst, usec_str);
            dst += 6;
            src += 2;
        } else {
            *dst++ = *src++;
        }
    }
    *dst = '\0';
    
    strftime(buf, buf_size, temp_format, tm_info);
}

/* Inject thermal data into JSON object */
static void inject_json(cJSON *json_obj, cJSON *thermal_data, const struct timeval *tv, const char *time_format) {
    /* Add timestamp */
    char timestamp[64];
    format_timestamp(timestamp, sizeof(timestamp), tv, time_format);
    cJSON_AddStringToObject(json_obj, "TIMESTAMP", timestamp);
    
    cJSON *sub_obj = cJSON_CreateObject();
    cJSON_AddItemToObject(json_obj, "THERMOCOUPLE", sub_obj);

    cJSON *item = thermal_data->child;
    while (item) {
        cJSON *copy = cJSON_Duplicate(item, 1);
        cJSON_AddItemToObject(sub_obj, item->string, copy);
        item = item->next;
    }
}

/* Run the bridge - spawn cmg-cli and inject thermal data */
int bridge_run(FuseBridge *bridge) {
    /* Initialize boards first (before forking) */
    if (bridge_init_boards(bridge) != 0) {
        fprintf(stderr, "Error: Failed to initialize thermal boards\n");
        return 1;
    }
    
    int pipefd[2];
    if (pipe(pipefd) == -1) {
        perror("pipe");
        return 1;
    }
    
    pid_t pid = fork();
    if (pid == -1) {
        perror("fork");
        return 1;
    }
    
    if (pid == 0) {
        /* Child process - execute cmg-cli */
        close(pipefd[0]); /* Close read end */
        dup2(pipefd[1], STDOUT_FILENO); /* Redirect stdout to pipe */
        close(pipefd[1]);
        
        /* Build command: stdbuf -oL -eL cmg-cli get <args> */
        char **cmd_args = (char**)malloc((bridge->arg_count + 6) * sizeof(char*));
        cmd_args[0] = "stdbuf";
        cmd_args[1] = "-oL";
        cmd_args[2] = "-eL";
        cmd_args[3] = "cmg-cli";
        cmd_args[4] = "get";
        
        for (int i = 0; i < bridge->arg_count; i++) {
            cmd_args[5 + i] = bridge->args[i];
        }
        cmd_args[5 + bridge->arg_count] = NULL;
        
        execvp("stdbuf", cmd_args);
        perror("execvp");
        exit(1);
    } else {
        /* Parent process - read JSON lines and inject thermal data */
        close(pipefd[1]); /* Close write end */
        
        FILE *fp = fdopen(pipefd[0], "r");
        if (!fp) {
            perror("fdopen");
            return 1;
        }
        
        /* Install signal handlers for graceful shutdown */
        signals_install_handlers();
        
        char line[4096];
        
        while (g_running && fgets(line, sizeof(line), fp)) {
            /* Remove trailing newline */
            size_t len = strlen(line);
            if (len > 0 && line[len - 1] == '\n') {
                line[len - 1] = '\0';
            }
            
            /* Skip empty lines */
            if (strlen(line) == 0) {
                printf("\n");
                fflush(stdout);
                continue;
            }
            
            /* Capture timestamp when data arrives */
            struct timeval tv;
            gettimeofday(&tv, NULL);
            
            /* Try to parse as JSON */
            cJSON *json_obj = cJSON_Parse(line);
            if (json_obj) {
                /* Get thermal data and inject */
                cJSON *thermal_data = get_thermal_data(bridge);
                inject_json(json_obj, thermal_data, &tv, bridge->time_format);
                
                char *output = cJSON_PrintUnformatted(json_obj);
                printf("%s\n", output);
                fflush(stdout);
                
                free(output);
                cJSON_Delete(json_obj);
                cJSON_Delete(thermal_data);
            } else {
                /* Not JSON - pass through unchanged */
                printf("%s\n", line);
                fflush(stdout);
            }
        }
        
        fclose(fp);
        
        /* Wait for child process */
        int status;
        waitpid(pid, &status, 0);
        
        if (WIFEXITED(status)) {
            return WEXITSTATUS(status);
        } else {
            return 1;
        }
    }
    
    return 0;
}

/* Command: fuse - Fuse thermal data into cmg-cli output */
int cmd_fuse(int argc, char **argv) {
    char *config_path = NULL;
    int address = -1;
    int channel = -1;
    char key[64] = "TEMP_FUSED";
    char tc_type[8] = "K";
    char time_format[64] = "%Y-%m-%dT%H:%M:%S.%f";  /* Default with microseconds */
    
    /* Find '--' separator */
    int separator_idx = -1;
    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--") == 0) {
            separator_idx = i;
            break;
        }
    }
    
    if (separator_idx == -1) {
        fprintf(stderr, "Error: No '--' separator found\n");
        fprintf(stderr, "Usage: thermo-cli fuse [OPTIONS] -- [cmg-cli arguments...]\n");
        fprintf(stderr, "\nOptions:\n");
        fprintf(stderr, "  -C, --config FILE      Path to YAML/JSON config file\n");
        fprintf(stderr, "  -a, --address NUM      Single mode: Board address\n");
        fprintf(stderr, "  -c, --channel NUM      Single mode: Channel index\n");
        fprintf(stderr, "  -k, --key NAME         Single mode: JSON key to inject [default: TEMP_FUSED]\n\n");
        fprintf(stderr, "  -t, --tc-type TYPE     Thermocouple type (default: K)\n");
        fprintf(stderr, "  -T, --time-format FMT  Timestamp format (default: %%Y-%%m-%%dT%%H:%%M:%%S.%%f)\n");
        fprintf(stderr, "                         Use %%f for 6-digit microseconds\n");
        fprintf(stderr, "\nNote: Data fusion only works with JSON output from cmg-cli.\n");
        fprintf(stderr, "      The --json flag will be added automatically if not specified.\n");
        fprintf(stderr, "\nExamples:\n");
        fprintf(stderr, "  thermo-cli fuse --address 0 --channel 1 --key MY_TEMP -- --power\n");
        fprintf(stderr, "  thermo-cli fuse --config config.yaml -- --actuator --stream 5\n");
        fprintf(stderr, "  thermo-cli fuse -a 0 -c 0 -T '%%H:%%M:%%S.%%f' -- --power\n");
        return 1;
    }
    
    /* Parse options before '--' */
    optind = 1; /* Reset getopt */
    static struct option long_options[] = {
        {"config", required_argument, 0, 'C'},
        {"address", required_argument, 0, 'a'},
        {"channel", required_argument, 0, 'c'},
        {"key", required_argument, 0, 'k'},
        {"tc-type", required_argument, 0, 't'},
        {"time-format", required_argument, 0, 'T'},
        {0, 0, 0, 0}
    };
    
    int opt;
    while (optind < separator_idx && 
           (opt = getopt_long(separator_idx, argv, "C:a:c:k:t:T:", long_options, NULL)) != -1) {
        switch (opt) {
            case 'C': config_path = optarg; break;
            case 'a': address = atoi(optarg); break;
            case 'c': channel = atoi(optarg); break;
            case 'k': strncpy(key, optarg, sizeof(key) - 1); break;
            case 't': strncpy(tc_type, optarg, sizeof(tc_type) - 1); break;
            case 'T': strncpy(time_format, optarg, sizeof(time_format) - 1); break;
            default:
                fprintf(stderr, "Usage: thermo-cli fuse [OPTIONS] -- [cmg-cli arguments...]\n");
                return 1;
        }
    }
    
    /* Prepare sources */
    Config config = {0};
    ThermalSource single_source = {0};
    ThermalSource *sources = NULL;
    int source_count = 0;
    
    if (config_path) {
        /* Load from config file */
        if (config_load(config_path, &config) != THERMO_SUCCESS) {
            fprintf(stderr, "Error loading config file\n");
            return 1;
        }
        sources = config.sources;
        source_count = config.source_count;
    } else if (address >= 0 && channel >= 0) {
        /* Single source from CLI args */
        strncpy(single_source.key, key, sizeof(single_source.key) - 1);
        single_source.address = address;
        single_source.channel = channel;
        strncpy(single_source.tc_type, tc_type, sizeof(single_source.tc_type) - 1);
        /* Initialize with default calibration and update interval */
        single_source.cal_coeffs.slope = DEFAULT_CALIBRATION_SLOPE;
        single_source.cal_coeffs.offset = DEFAULT_CALIBRATION_OFFSET;
        single_source.update_interval = DEFAULT_UPDATE_INTERVAL;
        sources = &single_source;
        source_count = 1;
    } else {
        fprintf(stderr, "Error: Must specify --config or (--address and --channel)\n");
        return 1;
    }
    
    /* Extract arguments after '--' */
    char **fuse_args = &argv[separator_idx + 1];
    int fuse_arg_count = argc - separator_idx - 1;
    
    if (fuse_arg_count == 0) {
        fprintf(stderr, "Error: No arguments provided after '--'\n");
        return 1;
    }
    
    /* Check if --json or -j is already present */
    int has_json_flag = 0;
    for (int i = 0; i < fuse_arg_count; i++) {
        if (strcmp(fuse_args[i], "--json") == 0 || strcmp(fuse_args[i], "-j") == 0) {
            has_json_flag = 1;
            break;
        }
    }
    
    /* Build final args, adding --json if needed */
    char **final_args = NULL;
    int final_arg_count = fuse_arg_count;
    
    if (!has_json_flag) {
        /* Add --json flag */
        final_arg_count = fuse_arg_count + 1;
        final_args = (char**)malloc(final_arg_count * sizeof(char*));
        for (int i = 0; i < fuse_arg_count; i++) {
            final_args[i] = fuse_args[i];
        }
        final_args[fuse_arg_count] = "--json";
    } else {
        final_args = fuse_args;
    }
    
    /* Create and run bridge */
    FuseBridge *bridge = bridge_create(sources, source_count, final_args, final_arg_count, time_format);
    int exit_code = bridge_run(bridge);
    bridge_free(bridge);
    
    /* Free allocated args array if we created one */
    if (!has_json_flag) {
        free(final_args);
    }
    
    if (config_path) {
        config_free(&config);
    }
    
    return exit_code;
}
